src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs

```diff
@@ -43,253 +54,197 @@ protected AbstractUseAutoPropertyAnalyzer()
         protected abstract SyntaxNode GetFieldNode(TFieldDeclaration fieldDeclaration, TVariableDeclarator variableDeclarator);
 
         protected abstract void RegisterIneligibleFieldsAction(
-            List<AnalysisResult> analysisResults, HashSet<IFieldSymbol> ineligibleFields,
-            Compilation compilation, CancellationToken cancellationToken);
-
-        protected sealed override void InitializeWorker(AnalysisContext context)
-            => context.RegisterSemanticModelAction(AnalyzeSemanticModel);
+            ConcurrentSet<IFieldSymbol> ineligibleFields, SemanticModel semanticModel, SyntaxNode codeBlock, CancellationToken cancellationToken);
 
-        private void AnalyzeSemanticModel(SemanticModelAnalysisContext context)
-        {
-            var cancellationToken = context.CancellationToken;
-            var semanticModel = context.SemanticModel;
+        protected abstract void RegisterNonConstructorFieldWrites(
+            ConcurrentDictionary<IFieldSymbol, ConcurrentSet<SyntaxNode>> fieldWrites, SemanticModel semanticModel, SyntaxNode codeBlock, CancellationToken cancellationToken);
 
-            // Don't even bother doing the analysis if the user doesn't even want auto-props.
-            var option = context.GetAnalyzerOptions().PreferAutoProperties;
-            if (!option.Value)
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSymbolStartAction(context =>
             {
-                return;
-            }
-
-            var analysisResults = new List<AnalysisResult>();
-            var ineligibleFields = new HashSet<IFieldSymbol>();
+                var namedType = (INamedTypeSymbol)context.Symbol;
+                if (namedType.TypeKind is not TypeKind.Class and not TypeKind.Struct and not TypeKind.Module)
+                    return;
 
-            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
-            AnalyzeCompilationUnit(context, root, analysisResults);
+                var analysisResults = new ConcurrentQueue<AnalysisResult>();
+                context.RegisterSyntaxNodeAction(context => AnalyzePropertyDeclaration(context, namedType, analysisResults), PropertyDeclarationKind);
 
-            RegisterIneligibleFieldsAction(
-                analysisResults, ineligibleFields,
-                semanticModel.Compilation, cancellationToken);
-            Process(analysisResults, ineligibleFields, context);
-        }
+                var ineligibleFields = new ConcurrentSet<IFieldSymbol>();
+                var nonConstructorFieldWrites = new ConcurrentDictionary<IFieldSymbol, ConcurrentSet<SyntaxNode>>();
```

Comment:
Do we have pooled ConcurrentSet and pooled ConcurrentDictionary utility types? This seems like a good place to use them if they exist.

Reply:
i'll add pools here.

Reply:
done.